File: chicken.py
# Solution to Atharva and Chicken
# by Jacob Magnuson

# Read number of testcases
T = int(input())

for t in range(T):
    # Read total number of days, skip days, and pickle probability
    inp = input().split()
    n, d, p = int(inp[0]), int(inp[1]), float(inp[2])

    '''
    We'll solve this problem with dynamic programming. In the array memo below,
    memo[x] will correspond to the answer if the semester is x days long.

    If there are x days left in the semester, and Atharva gets pickles, he won't
    return until there are x - d days left in the semester (unless there are
    fewer than d days remaining, in which case he won't return at all). If he
    doesn't get pickles, he will return when there are x - 1 days left. We can
    then compute his expected number of pickles by using the expected number on
    those two other days (x - d and x - 1) as well as the expected number of
    pickles he gets on day x specifically.

    When Atharva gets pickles, he equiprobably gets 1, 2, or 3 pickles, so he
    can expect to get the average (2) when he does get pickles. Furthermore, if
    Atharva gets pickles with probability p, he doesn't get them with
    probability (1 - p). Putting this all together, we can compute the expected
    number of pickles he gets on day x as follows:

        memo[x] = p * (2 + memo[x - d])     +   (1 - p) * (memo[x - 1])

    With probability p, he gets (an expected) 2 pickles, and returns when the
    semester has x - d days remaining. With probability (1 - p), he gets no
    pickles, and returns when there is 1 day remaining.
    '''
    memo = [0]
    for x in range(1, n + 1):
        skip = 0 if x <= d else memo[x - d]
        memo.append(p * (2 + skip) + (1 - p) * memo[-1])

    print(memo[-1])

File: chicken.cpp
/*
Atharva and Chicken Solution
Problem Author: Charles Bailey
Solution Author: Atharva Nagarkar

This problem can be solved with simple dynamic programming.
Our state will be dp[day], which represents the expected number of pickles on the "day"th day.
This state relies on two other values:
	1. if we don't get pickles, then we go forward one day like normal
	2. otherwise, we skip "d" days ahead
Thus, one can solve the problem starting from the last day. This way, all states that are needed to solve for a smaller day have already been solved.
*/

#include <bits/stdc++.h>

using namespace std;

const int N = (int) (2e5 + 100);

int t, n, d;
double p;
double dp[N];

int main() {
	cout.precision(16);
	cin >> t;
	while(t-->0) {
		cin >> n >> d >> p;
		fill(dp, dp + n + n, 0);
		for(int day = n - 1; day >= 0; --day) {
			double skip = (1 - p) * dp[day + 1]; //if we don't get pickles, we go to the next day; the probability of this happening is (1 - probability of getting pickles)
			double choose = 0; //otherwise, there is a 1/3 * p probability of getting 1, 2, or 3 pickles, which makes a total of p probability of getting any pickles at all.
			for(int i = 1; i <= 3; ++i) {
				choose += 1.0 / 3.0 * p * (i + dp[day + d]);  //this also takes us to the (day + d)'th day, which we have already computed
			}
			dp[day] = skip + choose; //simply add the two expected values to get the expected value for the current day
		}
		cout << dp[0] << endl;
	}
}